lib/evmone/eof.cpp

```diff
@@ -129,45 +410,105 @@ std::variant<EOF1Header, EOFValidationError> validate_eof1(
 
-    if (data_size == 0)
-        return 7;  // MAGIC + VERSION + SECTION_ID + SIZE + TERMINATOR
-    else
-        return 10;  // MAGIC + VERSION + SECTION_ID + SIZE + SECTION_ID + SIZE + TERMINATOR
+    for (size_t code_idx = 0; code_idx < header.code_sizes.size(); ++code_idx)
+    {
+        const auto error_instr = validate_instructions(rev, header.get_code(container, code_idx));
+        if (error_instr != EOFValidationError::success)
+            return error_instr;
+
+        if (!validate_rjump_destinations(header.get_code(container, code_idx)))
+            return EOFValidationError::invalid_rjump_destination;
+
+        auto msh_or_error =
+            validate_max_stack_height(header.get_code(container, code_idx), code_idx, header.types);
+        if (const auto* error = std::get_if<EOFValidationError>(&msh_or_error))
+            return *error;
+        if (std::get<int32_t>(msh_or_error) != header.types[code_idx].max_stack_height)
+            return EOFValidationError::invalid_max_stack_height;
+    }
+
+    return header;
 }
 
+}  // namespace
+
 bool is_eof_container(bytes_view container) noexcept
 {
     return container.size() > 1 && container[0] == MAGIC[0] && container[1] == MAGIC[1];
 }
 
-EOF1Header read_valid_eof1_header(bytes_view container) noexcept
+/// This function expects the prefix and version to be valid, as it ignores it.
+EOF1Header read_valid_eof1_header(bytes_view container)
 {
+    EOFSectionHeaders section_headers;
+    auto it = container.begin() + std::size(MAGIC) + 1;  // MAGIC + VERSION
+    while (*it != TERMINATOR)
+    {
+        const auto section_id = *it++;
+        if (section_id == CODE_SECTION)
+        {
+            const auto code_section_num = read_uint16_be(it);
+            it += 2;
+            for (uint16_t i = 0; i < code_section_num; ++i)
+            {
+                const auto section_size = read_uint16_be(it);
+                it += 2;
+                section_headers[section_id].emplace_back(section_size);
+            }
+        }
+        else
+        {
+            const auto section_size = read_uint16_be(it);
+            it += 2;
+            section_headers[section_id].emplace_back(section_size);
+        }
+    }
+    const auto header_size = eof_header_size(section_headers);
+
     EOF1Header header;
-    const auto code_size_offset = 4;  // MAGIC + VERSION + CODE_SECTION_ID
-    header.code_size =
-        static_cast<uint16_t>((container[code_size_offset] << 8) | container[code_size_offset + 1]);
-    if (container[code_size_offset + 2] == 2)  // is data section present
+
+    for (auto type_offset = header_size;
+         type_offset < header_size + section_headers[TYPE_SECTION][0]; type_offset += 4)
+    {
+        header.types.emplace_back(container[type_offset], container[type_offset + 1],
+            read_uint16_be(&container[type_offset + 2]));
+    }
+
+    header.code_sizes = section_headers[CODE_SECTION];
+    std::vector<uint16_t> code_offsets;
```

Comment:
Seems to be a stray variable now, because the code writes to` header.code_offsets` directly.